Refactor toMap mecanisms

cchantep opened the pull request.

cchantep commented on framework/src/play/src/main/scala/play/api/data/Form.scala:
```diff
@@ -38,18 +38,20 @@ case class Form[T](mapping: Mapping[T], data: Map[String, String], errors: Seq[F
-  val constraints: Map[String, Seq[(String, Seq[Any])]] = mapping.mappings.map { m =>
-    m.key -> m.constraints.collect { case Constraint(Some(name), args) => name -> args }
-  }.filterNot(_._2.isEmpty).toMap
+  val constraints: Map[String, Seq[(String, Seq[Any])]] =
```
Refactor steps `.map(.collect)`+`.filterNot`+`.toMap` with `.collect(.collect)`

cchantep commented on framework/src/play/src/main/scala/play/api/data/Form.scala:
```diff
@@ -38,18 +38,20 @@ case class Form[T](mapping: Mapping[T], data: Map[String, String], errors: Seq[F
   /**
    * Formats associated to this form, indexed by field name. *
    */
-  val formats: Map[String, (String, Seq[Any])] = mapping.mappings.map { m =>
```
Refactor `.map`+`.collect`+`.toMap` with `.flatMap`

cchantep commented on framework/src/play/src/main/scala/play/api/i18n/Messages.scala:
```diff
@@ -538,11 +539,11 @@ class DefaultMessagesApiProvider @Inject() (
     config.get[Boolean]("play.i18n.langCookieHttpOnly")
 
   protected def loadAllMessages: Map[String, Map[String, String]] = {
-    langs.availables.map(_.code).map { lang =>
```
Refactor `.map`+`.toMap` with `.map` (up to the first `+`)

cchantep commented on framework/src/play/src/main/scala/play/api/mvc/Cookie.scala:
```diff
@@ -505,11 +505,16 @@ trait UrlEncodedCookieDataCodec extends CookieDataCodec {
       if (data.nonEmpty) {
-        data
-          .split("&")
-          .map(_.split("=", 2))
```
Refactor `.map(.split)`+`.map`+`.toMap` with `.flatMap(.span+match)`

> Find `.span` safer than `.split=>Array.apply`

cchantep commented on framework/src/play/src/main/scala/play/api/mvc/Headers.scala:
```diff
@@ -54,9 +54,9 @@ class Headers(protected var _headers: Seq[(String, String)]) {
-  override def equals(other: Any) = {
-    other.isInstanceOf[Headers] &&
-      toMap == other.asInstanceOf[Headers].toMap
+  override def equals(that: Any) = that match {
```
Avoid `isInstanceOf`+`.asInstanceOf` using pattern matching

cchantep commented on framework/src/play/src/main/scala/play/api/mvc/Headers.scala:
```diff
@@ -98,12 +98,15 @@ class Headers(protected var _headers: Seq[(String, String)]) {
    */
   lazy val toMap: Map[String, Seq[String]] = {
-    val map = headers.groupBy(_._1.toLowerCase(Locale.ENGLISH)).map {
+    val builder = TreeMap.newBuilder[String, Seq[String]](CaseInsensitiveOrdered)
```
After `headers.groupBy`, refactor `.map`+`toSeq`+`TreeMap.apply(T*)` with `.foreach(buf+=)`+`buf.result()` using a local builder approach

cchantep commented on framework/src/play/src/main/scala/play/api/mvc/RangeResult.scala:
```diff
@@ -363,10 +363,17 @@ object RangeResult {
-      Some(ACCEPT_RANGES -> "bytes"),
-      fileName.map(f => CONTENT_DISPOSITION -> s"""attachment; ${HttpHeaderParameterEncoding.encode("filename", f)}""")
-    ).flatten.toMap
+    val commonHeaders = {
```
Refactor `Seq.apply(*)`+`.flatten`+`.toMap` with `buf+=`+`Option.foreach(buf+=)`+`.result()` using a builder

cchantep commented on framework/src/play/src/main/scala/play/api/routing/sird/QueryStringExtractors.scala:
```diff
@@ -28,15 +28,14 @@ trait QueryStringParameterExtractor[T] {
-    }.groupBy(_._1).mapValues(_.map(_._2)).toMap)
-      .getOrElse(Map.empty)
-  }
+  private def parse(query: String): QueryString =
```
For non-null `query` (inside), refactor `.split`+`.toSeq`+`.map(.split)`+`.groupBy`+`.mapValues(.toSeq.map.toMap)` with `.split`+`.map(.span+match)`+`.groupBy`+`.mapValues(.toSeq.map)`

At top/`Option(query)` level, also refactor `.map.getOrElse` with `.fold`

wsargent commented:
@cchantep how did you find all this?

cchantep commented:
`git grep toMap` in `src/play` module (not checked the other modules)

cchantep commented on framework/src/play/src/main/scala/play/api/http/HttpConfiguration.scala:
```diff
@@ -220,13 +220,20 @@ object HttpConfiguration {
-        .filter(_ (0) != '#')
-        .map(_.split('='))
-        .map(parts => parts(0) -> parts.drop(1).mkString).toMap
+        config.get[String]("play.http.fileMimeTypes").split('\n').flatMap { l =>
```
After `split('\n')`, refactor `.map(.trim)` + `.filter` + `.filter` + `map(.split)` + `.map` + `.toMap`, as `.flatMap(.trim.splitAt + match + span + match)`

cchantep commented on framework/src/play/src/main/scala/play/api/mvc/Cookie.scala:
```diff
@@ -504,13 +504,21 @@ trait UrlEncodedCookieDataCodec extends CookieDataCodec {
+        Map.empty[String, String]
+      } else {
+        data.split("&").flatMap { pair =>
+          pair.span(_ != '=') match { // "foo=bar".span(_ != '=') -> (foo,=bar)
```
Refactor `.split(.., 2)` with safer `.span`

cchantep replied:
`.span`+`match` makes sure there are at least 2 fragments (the previous `p(0)` and `p(1)` were possibly raising `NoSuchElementException` in the wild way, rather than ignoring invalid pairs).

There was not specs about the case `a=foo=bar&b=lorem` with the first pair `a` having the value `foo=bar`.

marcospereira commented on framework/src/play/src/main/scala/play/api/http/HttpConfiguration.scala:
```diff
@@ -220,13 +220,20 @@ object HttpConfiguration {
+            case ("#", _) => Option.empty[(String, String)] // comment
+
+            case _ => // "foo=bar".span(_ != '=') -> (foo,=bar)
+              line.span(_ != '=') match {
```
Why not using `split` here too? It would be simpler to read and you won't need `drop` below.

cchantep replied:
The first `.split` is ok as the expected result is a `Traversable` 0..n, but there the result is expected to be 2 fragments, with no safety to have even one using `.split` (which was raising `NoSuchElementException` which has to be caugh then).

marcospereira commented on framework/src/play/src/main/scala/play/core/parsers/Multipart.scala:
```diff
@@ -337,10 +340,13 @@ object Multipart {
+
+                  (key.trim.toLowerCase(java.util.Locale.ENGLISH), value.mkString(":").trim)
+
+                }.toMap
```
`breakOut` here too?

cchantep replied:
Not for `Iterator`

marcospereira commented on framework/src/play/src/test/scala/play/api/mvc/FlashCookieSpec.scala:
```diff
@@ -78,21 +78,25 @@ class FlashCookieSpec extends Specification {
-      val m = flash.decode(es)
-      m.size must_== 0
+
+      flash.decode(es).size must_== 0
```
`flash.decode(es) must haveSize(0)`.

cchantep commented:
@marcospereira everywhere the goal is to split in a fixed expectation of 2 parts, that's safer to use `.span` for me (previously the usage of `.split` in such case was hiding `NoSuchElementException`).  For me `.split` is ok when a result 0..n  is expected.

marcospereira commented on framework/src/play/src/test/scala/play/api/mvc/FlashCookieSpec.scala:
```diff
@@ -82,19 +82,19 @@ class FlashCookieSpec extends Specification {
       val es = oldEncoder(Map("a" -> "b", "c" -> "d"))
 
-      flash.decode(es).size must_== 0
+      flash.decode(es) must be empty
```
Build is failing. This needs to be`must beEmpty`.

cchantep commented:
Applied @gmethvin
